### ISSUE TITLE
Add extension points for speech to avoid NVDA Remote monkeypatching

### ISSUE THREAD

**Issue by LeonarddeR** [nvaccess/nvda#14520]
### Is your feature request related to a problem? Please describe.
NVDA Remote currently relies on a lot of monkeypatching to intercept braille, speech, gestures, tones and waves from the controlled system. #14503 intends to fix most of this, except for speech.

Speech is a difficult thing. Historically. NVDA Remote patched methods on the synth directly, e.g. speak, cancel. With speech refactor becoming available, this was changed to patch methods on the speech manager. This had several advantages, including:

1. Indexing and callback commands could be handled by the controller. This means that beeping in the middle of a sentence would work correctly, as the beep callback is executed by the controller
2. It was no longer necessary to patch the synthesizer

However, there are some drawbacks and current shortcomings:

1. NVDA Remote must mess around with the speech state. `speech.speech._speechState.beenCanceled` is set to False before NVDA remote can speak
2. Cancellable speech does not work

### Describe the solution you'd like
I think that for NVDA Remote, the current approach should be preferred over the older approach where the controller relies on controlled system indexing. In other words, shortcoming 2 does not outweigh advantage 1. To solve shortcoming 1, it might be necessary to use speech.speak, speech.cancelSPeech and speech.pauseSPeech directly. On the other hand, for #3564, it makes more sense to work with a virtual synth driver, since the controlled system (or server) doesn't require local speech output. In that case, it's fine if the controlled system does the indexing and cancellable speech should work as well.

That said, we should solve shortcoming 1, e.g. NVDA Remote should be able to queue speech without fiddling with the internal speech state. Now I'm not very comfortable with this new approach yet. Therefore I hope @feerrenrut can share his ideas about this.

To avoid monkey patching, we need several extension points to cover both approaches:

* synthDriverHandler.synthChanged: To perform actions when a new synth is selected
* An extension point for speak. This can also be used for the speechViewer
* An extension point vor cancel
* An extension point for pause.

Additional requirements:

* Consider moving the filtering of redundant LangChangeCommand objects and handling of character commands, character and speechdict processing from speech.speech.speak to speech.manager.SPeechManager._processSpeechSequence. Note that it would make sense if a controller can use own preferences for language switching, symbol level, etc.

I'd love thoughts from @feerrenrut, @ctoth and @tspivey as well as other interested people.

**Comment by ABuffEr** [comment]
Hi,
I'm following @leonardder 's work with high interest. I'm not sure if my observation is valid/important (I must see extension points in action), but should not this PR include speechDictHandler.processText?
I monkey-patched it in widely appreciated [NumberProcessing](https://github.com/ABuffEr/NumberProcessing), but trying to introduce continous mode in [Unicode-ASCII](https://github.com/sukiletxe/nvda-unicode-ascii/) by @sukiletxe, I realized that - for my knowledge - two add-ons of this kind cannot operate easily and safely at the same time. While it'd be possible with an extension point over speechDictHandler.processText, I suppose.

**Comment by leonardder** [comment]
> * [Instant translate](https://github.com/beqabeqa473/instantTranslate) maintained by @beqabeqa473

Why would instant translate need to patch speech.speak?

**Comment by leonardder** [comment]
@ABuffEr I guess speechDictHandler.processText could have a filter to accomplish this. Might be best to do this in a separate pr. There is now a filter in the braille module, so it should be pretty trivial to look at that and implement this for even a new dev.

**Comment by CyrilleB79** [comment]
> > * [Instant translate](https://github.com/beqabeqa473/instantTranslate) maintained by @beqabeqa473
> 
> Why would instant translate need to patch speech.speak?

Because there is a script to translate the last spoken text. The patch is used to store in a variable the last speech sequence passed to this function.

**Comment by jscholes** [comment]
> there is a script to translate the last spoken text

While the proposed extension point(s) would help with this use case, it also seems like a built-in NVDA speech history buffer would be more appropriate for such simple functionality.  I.e. allowing an add-on to pull the last spoken sequences on demand, rather than being pushed everything because one sequence might be needed at some point.

**Comment by seanbudd** [comment]
Now that the dust has settled, @leonardder do you plan on proceeding with this?

**Comment by leonardder** [comment]
I'm still not sure how to proceed regarding where to inject the extension point as well as how to deal with cancellable speech.

**Comment by beqabeqa473** [comment]
Yes, it would be great to see speech coming through extension point. That would eliminate lots of severe issues that would come from some addons incorrectly working with patching.

**Comment by seanbudd** [comment]
I think there is still some work left here right? @LeonarddeR could you update the issue with the remaining extension points requested?

**Comment by LeonarddeR** [comment]
I think the following points are still open:

* An extension point for pause.

Additional requirements:

* Consider moving the filtering of redundant LangChangeCommand objects and handling of character commands, character and speechdict processing from speech.speech.speak to speech.manager.SPeechManager._processSpeechSequence. Note that it would make sense if a controller can use own preferences for language switching, symbol level, etc.

**Comment by beqabeqa473** [comment]
Hi @LeonarddeR

If #16213 will be merged, will it be still needed to move processing of symbols and dictionaries to manager.speak? If we subscribe to an action, maybe it will be better to pass unfilterred speech to speech.speak on the controller side?

**Comment by beqabeqa473** [comment]
Yes, that is what i am saying that instead of patching manager.speak, we could subscribe to pre_filter_action and use speech.speak on the local system. Or maybe i am missunderstanding something?

**Comment by LeonarddeR** [comment]
I'm not sure whether that works instantly, but it makes sense for sure.

**Comment by Adriani90** [comment]
In #17428 speech pause and resume has been added.

So as far as I understand, following points are still open:
* Consider moving the filtering of redundant LangChangeCommand objects and handling of character commands, character and speechdict processing from speech.speech.speak to speech.manager.SPeechManager._processSpeechSequence.
* Use pre_speech and post_speech actions in speech viewer, code available in #17429.

**Comment by LeonarddeR** [comment]
I'm going to close this issue. The original goal was to avoid NVDA remote monkey patching, and remote is now in core. If there are any other extension points necessary, separate issues/pull requests can be filed.
